### PR TITLE
Allow usage with tensorflow 2.0.0 (via tf.compat.v1)

### PR DESCRIPTION
**Pull request by chriselion**
Still hacky but most unit tests are passing locally. Haven't tried any "real" training yet.

HUGE thank to @SetoKaiba for figuring out most of the issues in the branch mentioned in https://github.com/Unity-Technologies/ml-agents/issues/1850.

Main issues:
- [x] Better version checking 
- [x] cleanup some of the imports (I don't like tf_flatten, not sure mlagents.trainers is the right spot, etc).
- [x] Figure out what's up with tf2bc
- [x] Ban raw "import tensorflow" except in one location
- [x] Set up CI to work on 1.7, 1.14, and 2.0
- [x] Measure inference quality - quick test with 3dball didn't look good

### CODE REVIEW

File: ml-agents/mlagents/trainers/bc/models.py [hunk 1]
```diff
@@ -44,9 +44,7 @@ def __init__(
                         size,
                         activation=None,
                         use_bias=False,
-                        kernel_initializer=c_layers.variance_scaling_initializer(
-                            factor=0.01
-                        ),
+                        kernel_initializer=tf_variance_scaling(0.01),
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/bc/models.py]: As mentioned in the PR, I don't love the name. Would like to find a better way to handle this.

Also, the parameter name changed (factor vs scale) to I just left it as positional.

File: ml-agents/mlagents/trainers/tf.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import tensorflow as tf
+
+# TODO better version check, this will do for now though
+is_tf2 = tf.__version__ == "2.0.0"
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/tf.py]: There's definitely better ways to do this, just wanted something up and running initially.

**Comment by unityjeffrey** [comment]
Hi @chriselion, had a question.  I believe 2.0 does offer some backward compatibility to 1.x versions.  However, my understanding is, without making the traininer code 2.0 native, you don't get the major advantages of 2.0.  With that, when we say "2.0.0 support", does that mean we are going to (or need to) also make our code 2.0 native?

**Comment by chriselion** [comment]
@unityjeffrey - correct, the goal of this PR is not to take advantage of any 2.0 features. Currently if someone has tensorflow==2.0.0 installed, ml-agents will break (and it will complain at `pip install` time first). After this PR, we'll be compatible with 2.0.0

As for "2.0 native", I'll have to let other tf experts chime in. I doubt that there's a way to do that without breaking 1.x compat, so we'll have to decide if/when the new features are compelling enough to drop support for users on the old version.

**Comment by ervteng** [comment]
@unityjeffrey IMO TF 2.0 will make our trainer codebase a lot cleaner and more modular. However, as @chriselion says, it will completely break 1.0 compatibility, and it will take a substantial effort to change over.

**Comment by unityjeffrey** [comment]
ah i see.  makes sense.  would it be suffice to say that we upgrade to TF 2.0 now (but do not make it 2.0 ready).  But this will enable us later in the future to take on more work to make things cleaner and more modular (when we are ready)?

**Comment by ervteng** [comment]
> ah i see. makes sense. would it be suffice to say that we upgrade to TF 2.0 now (but do not make it 2.0 ready). But this will enable us later in the future to take on more work to make things cleaner and more modular (when we are ready)?

Yes

**Comment by chriselion** [comment]
@unityjeffrey Updated the PR title to "Allow usage with tensorflow 2.0.0 (via tf.compat.v1)", so hopefully that's less ambiguous about the goals.

File: ml-agents/mlagents/trainers/components/bc/model.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
-import tensorflow as tf
+from mlagents.trainers import tf
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/components/bc/model.py]: @ervteng @awjuliani @andrewcoh (research pod folks) How do you feel about importing this way instead (instead of `import tensorflow as tf`)? Would something like `import mlagents.tensorflow as tf` be nicer?

Note that either way, we should prevent directly importing "raw" tensorflow except for one central spot; I added some checks to flake8 to prevent this.

Comment (ervteng) [reply ml-agents/mlagents/trainers/components/bc/model.py]: Either should be OK. I kind of like the current way rather than the other one since it seems less like we have a "special" version of TensorFlow built into mlagents.

Comment (chriselion) [reply ml-agents/mlagents/trainers/components/bc/model.py]: By "current way", do you mean the one in the PR now, or on develop?

I feel like `mlagents.trainers` is the wrong place for it; maybe `mlagents.utils` or something like that?

Comment (ervteng) [reply ml-agents/mlagents/trainers/components/bc/model.py]: The one in the PR. 

There could definitely be an `mlagents.tf_utils` file - there are actually some other generic trainer functions that could go in there (e.g. `get_variable` in the SAC model, any checkpointing functionality, etc.).

Comment (chriselion) [reply ml-agents/mlagents/trainers/components/bc/model.py]: Moved the import wrapper to `mlagents.tf_utils`

**Comment by chriselion** [comment]
I did a full training run last week using tf 2.0, and ran some of the models for inference; behavior looked good.

File: ml-agents/mlagents/tf_utils/tf.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    tf_flatten = tf.layers.flatten
+    tf_rnn = tf.nn.rnn_cell
+
+    tf.disable_v2_behavior()
+else:
+    import tensorflow.contrib.layers as c_layers
+
+    tf_variance_scaling = c_layers.variance_scaling_initializer
+    tf_flatten = c_layers.flatten
+    tf_rnn = tf.contrib.rnn
```
Comment (ervteng) [inline ml-agents/mlagents/tf_utils/tf.py]: I actually think we might not need these contrib layers. The RNN and flatten even in TF 1.7 exist as `tf.layers.flatten` and `tf.nn.rnn` which is aliased to `tf.contrib` equivalents... I think

Comment (chriselion) [reply ml-agents/mlagents/tf_utils/tf.py]: Yeah, you're right. Not sure why I thought I needed those originally. I think we still need the "import tensorflow.compat.v1 as tf" though.
